### PR TITLE
Gui hookspecs

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR would combine #1849 and #1816 in one, handling the menubar adding for both. It's still WIP. In particular I want to namespace the menubar options.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
I've updated this following some nice work from @tlambert03 on https://github.com/napari/napari-plugin-engine/pull/10, so that the menubars are now nested plugin name followed by function/ class name for both dock widgets and function generated widgets. I quite like this now, but am thinking that we should probably prevent these widgets from appearing in the `Window` drop down too otherwise it's just too many places right?

The biggest thing I'm probably still uneasy about is that `Tuple[Callable, Dict, Dict]` for the `provide_function` hookspec. The first thing is the function, the second the magic_kwargs dict, and then the third are keyword arguments to `add_dock_widget`. I'm sort of tempted just to drop the third dict.

Maybe asking people to provide
```python
my_func.napari_menu_name
my_func.napari_area
my_func.napari_allowed_areas
```
directly on the function itself, like we do for the QWidget.

I might go ahead and implement that now to see how it feels. Curious how others feel? This is probably ready for some light review @tlambert03 @jni but it is still got some rough edges as mentioned above

**Comment by sofroniewn** [comment]
Forgot to include screenshot!

Screen Shot 2021-01-06 at 8 28 19 PM

File: napari/plugins/_builtins.py
Comment (jni) [inline napari/plugins/_builtins.py]: Is this gonna stick around?

File: napari/plugins/hook_specifications.py
Comment (jni) [inline napari/plugins/hook_specifications.py:332]: What does historic mean? I found [this explanation](https://github.com/napari/napari/issues/1272#issuecomment-633064212) by @tlambert03, together with a link to the [pluggy docs on it](https://pluggy.readthedocs.io/en/latest/index.html#historic-hooks), but it is still a mysterious term to me:

> (call historic means that the implementation will be called once immediately on registration, regardless of whether it is registered before or after the hook specification)

I am wondering whether we should rename it to e.g. `call_immediately=`, because `historic=` is clear as molasses. My brain goes to either deprecated hookspecs or hookspecs that provide undo behaviour.

**Comment by jni** [comment]
> Maybe asking people to provide
> 
> my_func.napari_menu_name
> my_func.napari_area
> my_func.napari_allowed_areas
> 
> directly on the function itself, like we do for the QWidget.

When I proposed something like this for scikit-image (function.extras), I asked Emma to ask Gaël what he thought, since he is somewhat of a stickler for simple APIs. He proposed shipping me a rubber duck with long curly hair that would respond "no." any time I asked about creating some exotic, non-standard API. 😂 

It was a good proposal, because now I imagine the gaël-duck every time this question comes up! 😂 

So, to be clear, I'm 👎 on this proposal, following the duck's advice.

All of the above "extras" are optional, right? In which case, I suggest replacing `Tuple[Callable, Dict, Dict]` with `Union[Callable, Tuple[Callable, Dict], Tuple[Callable, Dict, Dict]]` — would that work? Again, my medium/long term vision is that we come up with a set of type annotations that is descriptive enough that people *just* give us their type annotated functions and nothing else: the type annotations are enough for magicgui to do its magic, and simultaneously we come up with an awesome enough layout engine that allowed_areas is no longer needed.

As an example, we would something like:

```python
def gaussian(
    image : Image,
    sigma : Annotated[float, FloatRange(1, 100)],
) -> Image:
```

**Review by jni — COMMENTED** [review]
In addition to my API comment, I've made a couple of smaller comments on the current code.

**Comment by tlambert03** [comment]
> As an example, we would something like:
> ```python
> def gaussian(
>     image : Image,
>     sigma : Annotated[float, FloatRange(1, 100)],
> ) -> Image:
>  ```

This is already supported by magicgui, but you need to provide a dict:

```python
def gaussian(image, sigma: Annotated[float, {'min': 1, 'max': 100}]):
```

**Comment by tlambert03** [comment]
and tangentially related, just so people know... It's also possible to modify these annotations dynamically after the function definition, by modifying the `__annotations__` dict:
```python
In [1]: def f(x):
   ...:     pass
   ...:

In [2]: f.__annotations__
Out[2]: {}

In [3]: f.__annotations__['x'] = int

In [4]: f
Out[4]: <function __main__.f(x: int)>

In [5]: f?
Signature: f(x: int)
Docstring: <no docstring>
Type:      function
```
(so it would be possible for someone to provide a set of "widget annotations" for a third-party library like scikit-image)

**Comment by jni** [comment]
> This is already supported by magicgui, but you need to provide a dict:

Yeah I did get a sense of déjà vu but I couldn't tell if it was real 😅

But, then the extra dict should indeed be optional, right?

**Comment by tlambert03** [comment]
> > This is already supported by magicgui, but you need to provide a dict:
> 
> Yeah I did get a sense of déjà vu but I couldn't tell if it was real 😅
> 
> But, then the extra dict should indeed be optional, right?

Correct, both `int` and `Annotated[int, {...}]` are valid annotations,  and lacking the options, it will use magicgui’s default settings

(Edit: oops, if you mean the extra dict in the hookspec. Then not entirely, as there are some global widget magicgui options (auto_call, labels, layout, etc...) that aren’t captured in parameter annotations.  Thoughts on where to put that?)

File: napari/plugins/_builtins.py
Comment (sofroniewn) [reply napari/plugins/_builtins.py]: probably not, will drop

**Comment by sofroniewn** [comment]
> > As an example, we would something like:
> > ```python
> > def gaussian(
> >     image : Image,
> >     sigma : Annotated[float, FloatRange(1, 100)],
> > ) -> Image:
> > ```
> 
> This is already supported by magicgui, but you need to provide a dict:
> 
> ```python
> def gaussian(image, sigma: Annotated[float, {'min': 1, 'max': 100}]):
> ```

I wonder if we shouldn't start supporting all the pydantic custom types, for example their constrained types - see https://pydantic-docs.helpmanual.io/usage/types/#constrained-types

```python
from pydantic import confloat

def gaussian(image, sigma: confloat(ge=0, le=100)):
```

Their list is really quite comprehensive for generic types, and I'd prefer to reuse this system rather than invent our own. For example they have `confloat(ge=0, le=100)` and `confloat(gt=0, lt=100)` for both `<=` and `<` like inequalities. It's unclear what the default behavior with `float, {'min': 1, 'max': 100}` or `FloatRange(1, 100)` would be, and I think we'd just end up reinventing a lot of their system as we needed more complexity.

Note this doesn't have to be right now, but it's something we might want to consider if we start using pydantic for our base models.

**Comment by sofroniewn** [comment]
> So, to be clear, I'm 👎 on this proposal, following the duck's advice.
>
> All of the above "extras" are optional, right? In which case, I suggest replacing Tuple[Callable, Dict, Dict] with Union[Callable, Tuple[Callable, Dict], Tuple[Callable, Dict, Dict]] — would that work?

I can do that, but should we do that for the `QWidget` case too? @tlambert03 what are your thoughts here? I think I have the least strong preference about this, so will go with whatever people want.

File: napari/plugins/hook_specifications.py
Comment (sofroniewn) [reply napari/plugins/hook_specifications.py:332]: I think "historic" happens once and doesn't get changed, so I'm ok with the name. I don't have strong feelings, but maybe a mild preference to not deviating from pluggy names unless it is a last resort. I'll defer to @tlambert03. That would need to be a new breaking change for the napari plugin engine too, and not in this PR

**Comment by tlambert03** [comment]
just some thoughts ... haven't looked again at code yet:

> the menubars are now nested plugin name followed by function/ class name for both dock widgets and function generated widgets.

yeah this is great! From a ux perspective, I'm a little worried we're going to have a lot of subwindows with a single option (i.e. a plugin that only provides a single dock widget, which will probably be most of them?). Screen Shot 2021-01-07 at 4 43 53 PM
The Dock Widgets menu is great, but I'm wondering if we can't maybe avoid the submenus if the plugin only provides on widget? (To avoid naming conflicts, I'd prefer renaming the menu item to something like `QWidget (builtins)` as opposed to a "builtins" menu with 1 "QWidget").

> but am thinking that we should probably prevent these widgets from appearing in the Window drop down too otherwise it's just too many places right?

Yeah, totally agree. Also, at the moment, there's some desynchronization: if you close the widget in the Windows menu (or click the x button in the widget), it remains checked in the DockWidgets menu... but unchecking it in DockWidgets removes it from the other two.

> The biggest thing I'm probably still uneasy about is that Tuple[Callable, Dict, Dict] for the provide_function hookspec.

Yeah I agree with you, and I agree with @jni's comments that ultimately it'd be nice to _just_ have a well-annotated function. (which is also why I brought up the `Annotated` feature in euro-dev meeting the other day).  However, as @sofroniewn also pointed out during the meeting, there are rather useful some "widget-global" options (`call_button=True`, `labels=False, `auto_call=True`) that can't be tucked into an `Annotated` parameter type.  Add to that the need for people to provide some additional metadata for how this function should look in the menu, shortcuts, area restrictions, etc... and I think we're quite a ways away from the ideal of just being able to take a function.

I'm probably more in favor of an "overloaded single argument" approach like @jni's tuples (`Union[Callable, Tuple[Callable, Dict], Tuple[Callable, Dict, Dict]]`) than I am of adding attributes to the function like `my_func.napari_menu_name` 

> should we do that for the QWidget case too?

It feels very normal for a `class` which is already a natural "namespace object" and often has class attributes ... but it feels awkward for a function, which almost never has attributes like this.  So I don't "mind" the class attribute approach (lots of libraries use it). But if the symmetry of the `Union[...]` hookspec API was super appealing to anyone, I could be convinced either way.

> I wonder if we shouldn't start supporting all the pydantic custom types

I like that idea!  But I'd probably leave it forever (?) at "support", and not "require".  Happy to add some code to check if it's a pydantic constrained type, and convert it to some internal `Annotated[float, {...}]` type... but not looking to require any non-standard-lib stuff just to annotate a parameter.

File: napari/plugins/hook_specifications.py
Comment (tlambert03) [reply napari/plugins/hook_specifications.py:332]: i'm also ok with the name.  maybe wouldn't have picked it personally (and also wouldn't oppose a PR if someone really wanted to change it...), but it doesn't feel urgent to me.  The problem with `call_immediately` is that it's not really called immediately.  the `plugin_manager` gets to "call" the hookspec (at any point), and even if all the plugins haven't yet registered their implementations, when they _later_ get registered, they will then be able to provide results (like a list of widgets) for the hookspec that was called by the plugin manager in the past ("historically").

I feel like [this explanation](https://pluggy.readthedocs.io/en/latest/index.html#historic-hooks) from the pluggy docs mostly explains it?

> The implication is that late registered hookimpls will be called back immediately at register time ... This turns out to be particularly useful when dealing with lazy or dynamically loaded plugins.

Comment (tlambert03) [reply napari/plugins/hook_specifications.py:332]: > it's not really called immediately

oh ... I guess the plugin is called immediately :) not the hook call.  ah relativism

**Comment by jni** [comment]
> > I wonder if we shouldn't start supporting all the pydantic custom types
> 
> I like that idea! But I'd probably leave it forever (?) at "support", and not "require". Happy to add some code to check if it's a pydantic constrained type, and convert it to some internal Annotated[float, {...}] type... but not looking to require any non-standard-lib stuff just to annotate a parameter.

Agreed. Also, the pydantic stuff actually affects runtime behaviour of the function, clamping/raising on values outside their range, right? I don't think that's desirable. In the case of scikit-image, for example, I can see us using `Annotated`, but not `confloat`.

> But if the symmetry of the Union[...] hookspec API was super appealing to anyone, I could be convinced either way.

🙋

But, I agree that class attributes are also fine. I wouldn't mind supporting both in this case. But in any case I think a single widget class should be supported, just like a single callable should be supported.

> Then not entirely, as there are some global widget magicgui options (auto_call, labels, layout, etc...) that aren’t captured in parameter annotations.

Yes yes, we are agreed! I meant, "since we can get so much functionality out of Annotated, then I can imagine many use cases being met by just that, and we should *allow* the hookspec to provide *just* an annotated Callable. The dict *could* be provided for users that want additional options.

**Comment by tlambert03** [comment]
> Also, the pydantic stuff actually affects runtime behaviour of the function, clamping/raising on values outside their range, right? I don't think that's desirable.

Agreed, not desirable... but I don't think simply annotating a parameter as such would have any runtime effect, it would need to be the annotation for a field (class attribute) in a subclass of `pydantic.BaseModel` for it to have any runtime validation

**Comment by sofroniewn** [comment]
Ok changes pushed, this now includes a new menubar style, and I've dropped the additional function args that @jni didn't like, but I haven't added in the third dict argument to the hookspec. Maybe @jni can do that. This then needs a little cleaning up, but should be close.

File: napari/plugins/hook_specifications.py
Comment (sofroniewn) [inline napari/plugins/hook_specifications.py]: @tlambert03 we're getting fails from our headless mode check because of the qtpy imports. Should I change this to use strings and be forward reference, or do import inside `if TYPE_CHECKING`?

File: napari/plugins/__init__.py
Comment (sofroniewn) [inline napari/plugins/__init__.py]: Similar to comment bellow we're getting headless mode fails. Should I put imports in `if TYPE_CHECKING` or use string for annotations?

File: napari/plugins/hook_specifications.py
Comment (tlambert03) [reply napari/plugins/hook_specifications.py]: yep (both)!  import inside `if typing.TYPE_CHECKING` ... and then annotated with `Type['QWidget']`

Comment (tlambert03) [reply napari/plugins/hook_specifications.py]: (it will pass the tests with the forward reference alone, but mypy wouldn't work without the import inside of TYPE_CHECKING)

Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: ok, thanks I will give this a try now!

Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: ok, change pushed. let me know if that looks good

**Comment by sofroniewn** [comment]
Ok, made some updates, it needs some work inside `register_function` and `register_dock_widget` to account for all the different types of return values we now support, and needs some comprehensive tests. Getting close, but still needs a push

**Comment by sofroniewn** [comment]
Ok, following some suggestions from @tlambert03 this now has a `List` if people provide more than 1 thing, and a tuple if people are providing dicts. I don't do any validation, but i think we should (could be in future PR, could be now). I also have the type alias (?), but not sure if @jni will like that. I'm still missing tests and we need to write good doc strings.

I'd say with those caveats this is ready for review, and if @tlambert03/ @jni wanted to directly work on the type alias/ remove them or work on the doc strings then this would be much appreciated. I'm worried I'm not doing the best job possible here. Commits directly to the PR are welcome. I'll help with the tests.

I think this is pretty close now though!!! Which is very exciting :-)

**Comment by tlambert03** [comment]
> if @tlambert03/ @jni wanted to directly work on the type alias/ remove them or work on the doc strings then this would be much appreciated.

will review tonight.  I think the type aliases are fine (and I don't think @jni is globally against them?)  but in this case, I don't think I'd put them in `types.py`.  I know there are some other type aliases there, but that's just because we use them internally as type annotations throughout the repo.  In this case, _particularly_ to make it easier for someone just looking at at the hookspecs file, I think you should put the aliases next to the function.  Also, I wouldn't call it `MagicFunction` (the function is just the callable, so it's misleading to include a `Tuple` in the `Union`)  ... maybe just:
```python
ValidFunctionGuiArg = Union[Callable, Tuple[Callable, dict], Tuple[Callable, dict, dict]]
def hookspec() -> Union[ValidFunctionGuiArg, List[ValidFunctionGuiArg]]: ...
```

**Comment by tlambert03** [comment]
> that's just because we use them internally as type annotations throughout the repo

... oh I guess you do also use them in `plugins.__init__`, ah well, I don't know :) maybe keep 'em in `types`

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]: I like the behavior of having some menu item that clears all plugin dock widgets.  but I think the word "clear menu" is a little confusing, because it made me think I was just removing some items from the _menu_ (not that it would be closing widgets that I had been working on in the viewer).

Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: `Remove Dock Widgets` or `Clear Dock Widgets`?

File: napari/plugins/__init__.py
Comment (sofroniewn) [inline napari/plugins/__init__.py]:
```suggestion
        name = kwargs.get('name') or camel_to_spaces(_cls.__name__)
```
I think the `str` prevents the second thing from getting called. Was there a reason it was needed?

Comment (tlambert03) [reply napari/plugins/__init__.py]: ah right.  well, if for some reason they provide a non-string name, then it might cause an error when we use it as a menu item.  should use `str(kwargs.get('name', ''))`,  will change

Comment (sofroniewn) [reply napari/plugins/__init__.py]: thanks!! can you change in `register_function_widget` too

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: either `Remove Dock Widgets` or `Close All Dock Widgets`?

File: napari/plugins/hook_specifications.py
Comment (jni) [reply napari/plugins/hook_specifications.py:332]: > I feel like this explanation from the pluggy docs mostly explains it?

I still don't grok the link to the word "historic", but maybe I just need to brush up on my grammar? OED definition:

> (grammar) (of a tense) used in the narration of past events, especially Latin and Greek imperfect and pluperfect.

Agreed that this isn't urgent — this terminology doesn't leak into user/plugin-writer space, does it? In which case I'm happy to punt.

File: napari/plugins/hook_specifications.py
Comment (jni) [inline napari/plugins/hook_specifications.py]: What I dislike about aliases is that I have to hunt around for the definition, so I liked Talley's proposal of putting it next to the hookspec. If it needs to be used in multiple places, two options are:
- redefinition, which I think should work?
- including it just above the function in a comment.

One particular concern is how this will render. For example, if you look at our current hookspec docs:

https://napari.org/docs/dev/plugins/hook_specifications.html

you'll see some complicated-to-parse signatures *but* everything about those signatures is standard and easy to work through:

hookspec-type-annot

Do we want the docs to have only `MagicFunctionArg` there? Is there a standard tool that will expand that type alias on hover, or link aliases to their definition?

Given all this, I think probably the best way to do this is copy the alias in the docstring. Thoughts?

File: napari/plugins/hook_specifications.py
Comment (jni) [inline napari/plugins/hook_specifications.py]: This is out of date, btw. I think we can finalise it now?

**Review by jni — COMMENTED** [review]
@sofroniewn @tlambert03 I've left a comment about the type alias. If there's consensus on the fix I'm happy to push a change for it. Otherwise this looks good to me currently! :tada:

File: napari/plugins/hook_specifications.py
Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: This one is tricky.

I guess my ideal would be that `MagicFunctionArg` would be clickable in the docs and so you'd be able to click on it and then see what it is, and we'd only really have to define it in one place, and you wouldn't have to hunt around.

Without additional description something like
```python
Optional[Callable[[str, List[Tuple[Any, Dict, str]]], List[str]]]
```
is basically meaningless anyway. I don't think anyone could look at that and know what they are supposed to return without reading a lot more. I also struggle to parse all the nested `]` and so when I look at just have a really hard time with it.

I find the following much easier to parse
```python
Optional[Callable[[str, List[LayerDataTuple]], List[str]]]
```
It also makes it clear that `Tuple[Any, Dict, str]` in both `napari_get_writer` and `napari_get_reader` are the same thing.

Given by preference for the type aliases I'd probably just add the definition again in a comment above the code. I don't really like redefinition as they could then get out of sync.

File: napari/plugins/hook_specifications.py
Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: Yeah, I think we're just waiting to see if we go with `MagicFunctionArg` or not and how much to explain here

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: My top choice is probably `Remove Dock Widgets`. @jni?

File: napari/plugins/hook_specifications.py
Comment (jni) [reply napari/plugins/hook_specifications.py]: > Given by preference for the type aliases I'd probably just add the definition again in a comment above the code. I don't really like redefinition as they could then get out of sync.

Well, the nice thing is, if they get out of sync programmatically, our CI can catch it. If they get out of sync in the comments, then that is much harder to catch.

Also, adding the definition above the code doesn't solve the docs issue...

> I find the following much easier to parse
>
> Optional[Callable[[str, List[LayerDataTuple]], List[str]]]

A very good point! I'm on board that hover (click is not quite as friendly imho) on data type definitions is the ideal case. That might require some sphinx changes, but actually I'm seeing that they are either there or coming!

https://github.com/sphinx-doc/sphinx/pull/8007

A [GitHub code search for 'autodoc_type_aliases'](https://github.com/search?q=autodoc_type_aliases&type=code) took me here, where they seem to have correctly worked out how to link type aliases to their definition

https://pycgdescent.readthedocs.io/en/latest/#pycgdescent.minimize

The cool thing is that the definition includes a comment about how the alias works! 🤯

We should also alias something like `Pathlike = Union[str, pathlib.Path]`, I think.

One more thing, about the name: MagicFunctionArg seems vague, whereas LayerDataTuple feels concrete. How about AnnotatedFunction, GuiAnnotatedFunction, FunctionPlusGui, ...? Any other thoughts on this? Since this *is* now user-facing, I'd like to spend just a little bit of time coming up with a good name for it.

Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: > One more thing, about the name: MagicFunctionArg seems vague, whereas LayerDataTuple feels concrete. How about AnnotatedFunction, GuiAnnotatedFunction, FunctionPlusGui, ...? Any other thoughts on this? Since this is now user-facing, I'd like to spend just a little bit of time coming up with a good name for it.

Those are all possibilities, `AnnotatedFunction` might be my favourite, but I don't have strong feelings. We should think about what we do for the widget too. Could do `DockFunction` and `DockWidget` or `DockFunctionWidget` and `DockWidget`?

Comment (tlambert03) [reply napari/plugins/hook_specifications.py]: > One particular concern is how this will render. For example, if you look at our current hookspec docs you'll see some complicated-to-parse signatures but everything about those signatures is standard and easy to work through:

this remains a challenge with sphinx.  Type aliases (as opposed to `NewType`) really are currently just internal shorteners for the things they alias (though @jni ... https://github.com/sphinx-doc/sphinx/pull/8007 looks good!). I have used [https://github.com/agronholm/sphinx-autodoc-typehints](sphinx-autodoc-typehints) to at least make the signatures look better in docs (leaving the verbose explanation to the docstring).  Since our docs are still very much in flux, maybe it's not critical for this PR how it looks in the docs?

> Since this is now user-facing, I'd like to spend just a little bit of time coming up with a good name for it.

I agree, names are important.  worth getting right.  I think the function itself should just be `Callable`.  The `Union` between that callable and the `Tuple` with additional `dicts` is really what we need to decide.  I picked `...Arg` cause those are all valid "arguments" for our hook registration process.  I'm not sure anything with the word `Function` in it is appropriate for something where a `tuple` is also valid.

File: napari/plugins/hook_specifications.py
Comment (sofroniewn) [reply napari/plugins/hook_specifications.py:332]: > this terminology doesn't leak into user/plugin-writer space, does it?

I don't think it does. It's all internal to napari

File: napari/plugins/hook_specifications.py
Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: > Since our docs are still very much in flux, maybe it's not critical for this PR how it looks in the docs?

Yup, agreed.

> I agree, names are important. worth getting right. 

Yup. Logic for `...Arg` was sound to me. How about `DockFunctionArgs` and `DockWidgetArgs` for our two things respectively (I might lean towards `Args` a little over `Arg`, but would be fine with `Arg` if people thought that was more accurate).

Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: Or actually, even better, let's just go with the words in the hook spec itself, so

```python
def napari_experimental_provide_function_widget() -> Union[
    FunctionWidgetArgs, List[FunctionWidgetArgs]
]:
```
and
```python
def napari_experimental_provide_dock_widget() -> Union[
    DockWidgetArgs, List[DockWidgetArgs]
]:
```

Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: singular might be better than plural too, I don't mind

Comment (jni) [reply napari/plugins/hook_specifications.py]: See here for discussion: https://napari.zulipchat.com/#narrow/stream/270578-reviews/topic/2080/near/222245561 (requires registration/login with github). Summary:

- choice is not that critical — turns out to effectively only be a documentation choice, which can be reversed later. As such, this choice should not hold up merging of this PR.
- plural Args is better than singular Arg, as it more closely reflects that there is some unpacking to be done (even though it is not standard unpacking)
- all three of us gave the ok to FunctionWidgetArgs/DockWidgetArgs, *or* AnnotatedFunction/AnnotatedWidget, and all three of us liked a new option, AugmentedFunction/AugmentedWidget, which avoids the connotation/ambiguity as to whether these are type annotations or something else (or both).

This can get merged with or without the name change to `Augmented...`

**Comment by jni** [comment]
@sofroniewn @tlambert03 see https://github.com/sofroniewn/napari/pull/8

Still to do:
- tests for gui elements
- .. class directives to describe the typedefs

Maybe these can come in followup PRs, though...

**Comment by sofroniewn** [comment]
> @sofroniewn @tlambert03 see sofroniewn#8

just reviewed

> Maybe these can come in followup PRs, though...

i'm happy to give this one another 24hrs. I think we can have it wrapped up by then, but good to know this now has green light from you @jni.

@tlambert03 and I can put some finishing touches on it maybe during our day tomorrow and then merge.

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: renamed `Remove Dock Widgets`

File: napari/plugins/hook_specifications.py
Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: Renamed

File: napari/plugins/hook_specifications.py
Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: improved

File: napari/plugins/hook_specifications.py
Comment (sofroniewn) [inline napari/plugins/hook_specifications.py:338]: @jni language added here and below about experimental hookspec

**Comment by tlambert03** [comment]
more tests added @sofroniewn ... I think this more or less covers it

**Comment by sofroniewn** [comment]
Ok, great. This looks done to me! I've marked this as ready to review, maybe @tlambert03 @jni you can approve it and I will merge!! Thanks for all the hard work, this is a huge milestone for the project!!!

**Review by tlambert03 — APPROVED** [review]
🎉

**Pull request merged**
